rust/rsmgp-sys/src/vertex/mod.rs

```diff
@@ -0,0 +1,194 @@
+use c_str_macro::c_str;
+use std::ffi::{CStr, CString};
+
+use crate::context::*;
+use crate::edge::*;
+use crate::mgp::*;
+use crate::property::*;
+use crate::result::*;
+use crate::value::*;
+// Required here, if not present tests linking fails.
+#[double]
+use crate::mgp::ffi;
+use mockall_double::double;
+
+pub struct VerticesIterator {
+    ptr: *mut mgp_vertices_iterator,
+    is_first: bool,
+    context: Memgraph,
+}
+
+impl Default for VerticesIterator {
+    fn default() -> Self {
+        Self {
+            ptr: std::ptr::null_mut(),
+            is_first: true,
+            context: Memgraph {
+                ..Default::default()
+            },
+        }
+    }
+}
+
+impl Drop for VerticesIterator {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.ptr.is_null() {
+                ffi::mgp_vertices_iterator_destroy(self.ptr);
+            }
+        }
+    }
+}
+
+impl Iterator for VerticesIterator {
+    type Item = Vertex;
+
+    fn next(&mut self) -> Option<Vertex> {
+        unsafe {
+            let data: *const mgp_vertex;
+            if self.is_first {
+                self.is_first = false;
+                data = ffi::mgp_vertices_iterator_get(self.ptr);
+            } else {
+                data = ffi::mgp_vertices_iterator_next(self.ptr);
+            }
+
+            if data.is_null() {
+                None
+            } else {
+                let copy_ptr = ffi::mgp_vertex_copy(data, self.context.memory());
+                if copy_ptr.is_null() {
+                    panic!("Unable to allocate new vertex during vertex iteration.");
+                }
+                Some(Vertex {
+                    ptr: copy_ptr,
+                    context: self.context.clone(),
```

Comment:
Why cloning here? If context is global and unique?

Reply:
Memgraph (context) is actually not global, because each procedure call has a different context. `context::Memgraph` contains [std::rc::Rc](https://doc.rust-lang.org/std/rc/struct.Rc.html) to `context::MgpMemgraph`, which means there can be `&context::Memgraph` and copies of `context::Memgraph`. In this case, the copy is used, `&` is not convenient because then the lifetime has to be managed (Vertex with the `&` goes out of scope). The current implementation of `context::Memgraph` is the best I can currently think of. Open for suggestions on how to do this better :smile: